From feacd205bdce5ea6d3da942f88375b2fd4e9d632 Mon Sep 17 00:00:00 2001
From: Charles Ci-Jyun Wu <dminus@andestech.com>
Date: Wed, 31 May 2023 10:35:21 +0800
Subject: [PATCH 060/191] soc: andes: dma-noncoherent: select
 ARCH_HAS_DMA_CLEAR_UNCACHED to support driver free dma

Signed-off-by: Charles Ci-Jyun Wu <dminus@andestech.com>
---
 drivers/soc/andes/Kconfig | 1 +
 kernel/dma/direct.c       | 2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/soc/andes/Kconfig b/drivers/soc/andes/Kconfig
index 69147505c4b8..50eeb678d7d4 100644
--- a/drivers/soc/andes/Kconfig
+++ b/drivers/soc/andes/Kconfig
@@ -16,6 +16,7 @@ config ANDES_DMA_NONCOHERENT
 	select ARCH_HAS_SETUP_DMA_OPS
 	select ARCH_HAS_DMA_PREP_COHERENT
 	select ARCH_HAS_DMA_SET_UNCACHED
+	select ARCH_HAS_DMA_CLEAR_UNCACHED
 	depends on RISCV && PLAT_AE350
 	help
 	  Andes DMA noncoherent support
diff --git a/kernel/dma/direct.c b/kernel/dma/direct.c
index 63859a101ed8..8e6f7ea9dcf6 100644
--- a/kernel/dma/direct.c
+++ b/kernel/dma/direct.c
@@ -352,7 +352,7 @@ void dma_direct_free(struct device *dev, size_t size,
 	    dma_free_from_pool(dev, cpu_addr, PAGE_ALIGN(size)))
 		return;
 
-	if (is_vmalloc_addr(cpu_addr)) {
+	if (IS_ENABLED(CONFIG_DMA_DIRECT_REMAP) && is_vmalloc_addr(cpu_addr)) {
 		vunmap(cpu_addr);
 	} else {
 		if (IS_ENABLED(CONFIG_ARCH_HAS_DMA_CLEAR_UNCACHED))
-- 
2.34.1

